## Final Project - Submission

This documentation provides a comprehensive overview of the development and implementation of a neural network model for traffic sign classification.


In [1]:
# imports
import sys
 
sys.path.append("..")
 
from pathlib import Path
import mlflow
import torch.nn as nn
import matplotlib.pyplot as plt
 
from eval import evaluate, load_and_transform_data, get_sign_names, get_monte_carlo_predictions, SafetyBatchDataset
 
ROOT_DIR = Path().cwd().parent
BATCH_DIR = ROOT_DIR / "safetyBatches" / "Batch_5"
 
# Set the tracking server to be localhost with sqlite as tracking store
mlflow.set_tracking_uri(uri=f"sqlite:///{ROOT_DIR / 'mlruns.db'}")
 
# TODO: make sure you use the MLFlow run ID of your trained model
RUN_ID = "3737831d8c5e4f2c976b6bbde4ea97c0"     # with transformation
# RUN_ID = "633d2ccbf97943a3b9ed0a820c5baca1"   # without transformation
 
model_uri = f"runs:/{RUN_ID}/model"
loaded_model = mlflow.pytorch.load_model(model_uri)

# newcommand Monte-Carlo-Dropout
# num_forward_passes = 50
# num_classes = 43

# mean_value, variance_value = get_monte_carlo_predictions(
#     batch_loader, num_forward_passes, loaded_model, num_classes, len(batch_loader.dataset)
# )

criterion = nn.CrossEntropyLoss()
 
batch_loader = load_and_transform_data(data_directory_path=str(BATCH_DIR))
 
# newcommand
mlflow.end_run()
 
# newcommand
with mlflow.start_run(run_id=RUN_ID):
 
# TODO: track evaluation results
    predictions = evaluate(loaded_model, criterion, batch_loader, BATCH_DIR.name)
 
    # Output incorrect classifications
    ground_truth = []
    for _, target in batch_loader:
        ground_truth.extend(target.tolist())
    sign_names = get_sign_names()
    wrong_predictions_idx = [
        idx for idx, (y_pred, y) in enumerate(zip(predictions, ground_truth)) if y_pred != y
    ]
    for idx in wrong_predictions_idx:
        # image, target = batch_loader.dataset[idx]
        # plt.figure()
        # plt.imshow(image.permute(1, 2, 0), interpolation='bicubic')
        # plt.show()
        print(
            f"Traffic sign {sign_names[ground_truth[idx]]} incorrectly classified as {sign_names[predictions[idx]]}"
        )

2024/06/14 19:21:43 WARNING mlflow.pytorch: Stored model version '2.1.2+cpu' does not match installed PyTorch version '2.2.2+cpu'


Safety batch: Average loss: 0.0202, Accuracy: 76.1 %, Precision: 27.3 %
Traffic sign Speed limit (20km/h) incorrectly classified as Speed limit (30km/h)
Traffic sign Speed limit (20km/h) incorrectly classified as Speed limit (50km/h)
Traffic sign Speed limit (20km/h) incorrectly classified as Speed limit (50km/h)
Traffic sign Speed limit (20km/h) incorrectly classified as Speed limit (30km/h)
Traffic sign Speed limit (20km/h) incorrectly classified as Speed limit (30km/h)
Traffic sign Speed limit (30km/h) incorrectly classified as Road narrows on the right
Traffic sign Speed limit (30km/h) incorrectly classified as Dangerous curve to the right
Traffic sign Speed limit (30km/h) incorrectly classified as Road work
Traffic sign Speed limit (30km/h) incorrectly classified as Yield
Traffic sign Speed limit (30km/h) incorrectly classified as Dangerous curve to the left
Traffic sign Speed limit (30km/h) incorrectly classified as Dangerous curve to the left
Traffic sign Speed limit (30km/h) in

# Determination of robustness:

First try without data augmentation: 'Safety batch: Average loss: 0.0202, Accuracy: 76.1 %, Precision: 27.3 %'

Correctly classified images without transformation: 446 images

Accuracy after adding the transformation: 14.1%

Now we calculate <math xmlns="http://www.w3.org/1998/Math/MathML">
  <mrow>
    <mn>0.141</mn>
    <mo>&#x00D7;</mo>
    <mn>446</mn>
    <mo>=</mo>
    <mn>62.886</mn>
  </mrow>
</math>
to get the number of correctly classified images after the transformation. Here: 63 images

Our robustness value is calculated as <math xmlns="http://www.w3.org/1998/Math/MathML">
  <mrow>
    <mn>63</mn>
    <mo>&#x00F7;</mo>
    <mn>552</mn>
  </mrow>
</math>
and is 0.114 or 11.4%

---------------------------------------------

# Contact Information

- **Name:** Kaan Aydin

- **Course:** TFE21-2

- **Company:** ZF Friedrichshafen AG

- **Director of Studies:** Prof. Dr.-Ing. Thomas Kibler

- **Mail:** [kaan.aydin@zf.com](mailto:DeineEmail@example.com) 

- **University Mail:** [aydin.kaan-it21@it.dhbw-ravensburg.de](mailto:DeineEmail@example.com)

- <a href="https://www.linkedin.com/in/kaan-ayd/"><img src="https://upload.wikimedia.org/wikipedia/commons/e/e9/Linkedin_icon.svg" alt="LinkedIn" style="width:20px;height:20px;"></a> Kaan Aydin - LinkedIn

- <a href="https://github.com/KaanAyd"><img src="https://upload.wikimedia.org/wikipedia/commons/9/91/Octicons-mark-github.svg" alt="GitHub" style="width:20px;height:20px;"></a> Kaan Aydin - GitHub